OpenRA.Mods.Common/Activities/Resupply.cs

```diff
@@ -17,48 +17,94 @@
+	public class Resupply : Activity
 	{
 		readonly IHealth health;
 		readonly RepairsUnits[] allRepairsUnits;
 		readonly Target host;
 		readonly WDist closeEnough;
 		readonly Repairable repairable;
+		readonly Rearmable rearmable;
+		readonly INotifyResupply[] notifyResupplies;
 
 		int remainingTicks;
-		bool played = false;
+		bool played;
+		bool paused;
+		bool repairComplete;
+		bool rearmComplete;
+		ResupplyType activeResupplyTypes = ResupplyType.None;
 
-		public Repair(Actor self, Actor host, WDist closeEnough)
+		public Resupply(Actor self, Actor host, WDist closeEnough)
 		{
 			this.host = Target.FromActor(host);
 			this.closeEnough = closeEnough;
 			allRepairsUnits = host.TraitsImplementing<RepairsUnits>().ToArray();
 			health = self.TraitOrDefault<IHealth>();
 			repairable = self.TraitOrDefault<Repairable>();
+			rearmable = self.TraitOrDefault<Rearmable>();
+			notifyResupplies = host.TraitsImplementing<INotifyResupply>().ToArray();
+			repairComplete = health == null || health.DamageState == DamageState.Undamaged
+				|| repairable == null || !repairable.Info.RepairActors.Contains(host.Info.Name) || !allRepairsUnits.Any();
+
+			if (!repairComplete)
+				activeResupplyTypes |= ResupplyType.Repair;
+
+			rearmComplete = rearmable == null || !rearmable.Info.RearmActors.Contains(host.Info.Name) || rearmable.RearmableAmmoPools.All(p => p.FullAmmo());
+			if (!rearmComplete)
+				activeResupplyTypes |= ResupplyType.Rearm;
 		}
 
 		protected override void OnFirstRun(Actor self)
 		{
-			if (host.Actor.IsDead)
+			if (host.Type == TargetType.Invalid)
 				return;
 
-			foreach (var depot in host.Actor.TraitsImplementing<INotifyRepair>())
-				depot.BeforeRepair(host.Actor, self);
+			if (!repairComplete || !rearmComplete)
+				foreach (var notifyResupply in notifyResupplies)
+					notifyResupply.BeforeResupply(host.Actor, self, activeResupplyTypes);
+
+			// Reset the ReloadDelay to avoid any issues with early cancellation
+			// from previous reload attempts (explicit order, host building died, etc).
+			 // HACK: this really shouldn't be managed from here
+			if (!rearmComplete)
+				foreach (var pool in rearmable.RearmableAmmoPools)
+					pool.RemainingTicks = pool.Info.ReloadDelay;
 		}
 
 		public override Activity Tick(Actor self)
 		{
 			if (IsCanceling)
-			{
-				if (remainingTicks-- == 0)
-					return NextActivity;
+				return NextActivity;
+
+			if (host.Type == TargetType.Invalid || health == null)
+				return NextActivity;
+
+			if (closeEnough.LengthSquared > 0 && !host.IsInRange(self.CenterPosition, closeEnough))
+				return NextActivity;
 
-				return this;
+			if (!repairComplete)
+				RepairTick(self);
+
+			if (!rearmComplete)
+				RearmTick(self);
+
+			foreach (var notifyResupply in notifyResupplies)
+				notifyResupply.ResupplyTick(host.Actor, self, activeResupplyTypes);
+
+			if (repairComplete && rearmComplete)
+			{
+				Game.Debug("activeResupplyTypes: {0}", activeResupplyTypes);
```

Comment:
I guess that should be removed before merging.

Reply:
Oops, overlooked that.